utils/frame/remote-externalities/src/rpc_api.rs

```diff
@@ -18,83 +18,137 @@
 }
 
-/// Get the signed block identified by `at`.
-pub async fn get_block<Block, S>(from: S, at: Block::Hash) -> Result<Block, String>
-where
-	S: AsRef<str>,
-	Block: BlockT + serde::de::DeserializeOwned,
-	Block::Header: HeaderT,
-{
-	let client = build_client(from).await?;
-	let signed_block = client
-		.request::<SignedBlock<Block>>("chain_getBlock", rpc_params!(at))
-		.await
-		.map_err(|e| format!("chain_getBlock request failed: {:?}", e))?;
-
-	Ok(signed_block.block)
+/// Simple RPC service that is capable of keeping the connection.
+///
+/// Service will connect to `uri` for the first time during the first request. Instantiation
+/// does not trigger connecting.
+///
+/// Be careful with reusing the connection in a multithreaded environment.
+pub struct RpcService {
+	uri: String,
+	client: Arc<Mutex<Option<Client>>>,
+	keep_connection: bool,
 }
 
-/// Build a websocket client that connects to `from`.
-async fn build_client<S: AsRef<str>>(from: S) -> Result<WsClient, String> {
-	WsClientBuilder::default()
-		.max_request_body_size(u32::MAX)
-		.build(from.as_ref())
-		.await
-		.map_err(|e| format!("`WsClientBuilder` failed to build: {:?}", e))
-}
+impl RpcService {
+	/// Creates a new RPC service.
+	///
+	/// Does not connect yet.
+	pub fn new<S: AsRef<str>>(uri: S, keep_connection: bool) -> Self {
+		Self { uri: uri.as_ref().to_string(), client: Arc::new(Mutex::new(None)), keep_connection }
+	}
 
-/// Get the runtime version of a given chain.
-pub async fn get_runtime_version<Block, S>(
-	from: S,
-	at: Option<Block::Hash>,
-) -> Result<sp_version::RuntimeVersion, String>
-where
-	S: AsRef<str>,
-	Block: BlockT + serde::de::DeserializeOwned,
-	Block::Header: HeaderT,
-{
-	let client = build_client(from).await?;
-	client
-		.request::<sp_version::RuntimeVersion>("state_getRuntimeVersion", rpc_params!(at))
-		.await
-		.map_err(|e| format!("state_getRuntimeVersion request failed: {:?}", e))
+	/// Returns the address at which requests are sent.
+	pub fn uri(&self) -> String {
+		self.uri.clone()
+	}
+
+	/// Whether to keep and reuse a single connection.
+	pub fn keep_connection(&self) -> bool {
+		self.keep_connection
+	}
+
+	/// Build a websocket client that connects to `self.uri`.
+	async fn build_client(&self) -> Result<WsClient, String> {
+		WsClientBuilder::default()
+			.max_request_body_size(u32::MAX)
+			.build(&self.uri)
+			.await
+			.map_err(|e| format!("`WsClientBuilder` failed to build: {:?}", e))
+	}
+
+	/// Generic method for making RPC requests.
+	async fn make_request<'a, T: DeserializeOwned>(
+		&self,
+		call: RpcCall,
+		params: Option<ParamsSer<'a>>,
+	) -> Result<T, String> {
+		let mut maybe_client = self.client.lock().await;
+		match *maybe_client {
+			// `self.keep_connection` must be `true.
+			Some(ref client) => make_request(client, call, params).await,
+			None => {
+				let client = self.build_client().await?;
+				let result = make_request(&client, call, params).await;
+				if self.keep_connection {
```

Comment:
it would be much cleaner if you could create the connection in the constructor instead i.e, in `fn new` if `keep_connection == true` instead of having this mutex stuff.

I don't see any benefit of doing it this way, is there?

Reply:
laziness was my only excuse

Reply:
I mean, lazy creating of the connection, not mine :P

Reply:
Yes, it will help because you don't need to mutate the `RpcService` itself just instantiate the client in an `Arc` and because the `client` doesn't have to mutated anymore and thus you don't need a `Mutex` but you need to modify the constructor itself to be `async fn` which should be fine

Reply:
you could also have an enum instead of the Option

Reply:
in this case `Option` seems appropriate for me; I'm working on the change you suggested

Reply:
oh, only now I realized that you wrote `Option<Arc<..>>` instead of `Arc<Option<..>>`, which is ofc brilliant and indeed, looks better as an enum

Reply:
@niklasad1 I've applied your suggestion - it simplifies the stuff indeed, thank you

Reply:
looks good :)